File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.java
class Solution {
    public int[] validSequence(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();

        int [] rightHandSideMatchLength = new int[n + 1];

        int rightMatched = 0;
        int i = n - 1;
        int j = m - 1; 

        while(i >= 0){
            if(j >= 0 && word1.charAt(i) == word2.charAt(j)){
                rightMatched++;
                j --;
            }
            rightHandSideMatchLength[i] = rightMatched;
            i--;
        }

        List<Integer> seq = new ArrayList<>();
        boolean changePower = true;

        i = 0;
        j = 0;

        while(i < n && j < m){
            if (word1.charAt(i) == word2.charAt(j)){
                seq.add(i);
                j ++;
            }
            else if(changePower == true && rightHandSideMatchLength[i + 1] >= m - j - 1){
                seq.add(i);
                j ++;
                changePower = false;
            }
            i ++;
        }
        if(j == m){
            return seq.stream().mapToInt(Integer::intValue).toArray();
        } 
       return new int[0];
    }
}

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> list[int]:

        n = len(word1)
        m = len(word2)

        rightHandSideMatchLength = [0] * (n + 1)

        rightMatched = 0
        i = n - 1
        j = m - 1

        while i >= 0:

            if j >= 0 and word1[i] == word2[j]:
                rightMatched += 1
                j -= 1

            rightHandSideMatchLength[i] = rightMatched
            i -= 1

        seq = []

        changePower = True

        i = 0
        j = 0

        while i < n and j < m:

            if word1[i] == word2[j]:
                seq.append(i)
                j += 1

            elif changePower and rightHandSideMatchLength[i + 1] >= m - j - 1:
                seq.append(i)
                j += 1
                changePower = False

            i += 1

        if j == m:
            return seq

        return []